ic-factory/src/state.rs

```diff
@@ -1,139 +1,455 @@
     }
-}
 
-#[derive(CandidType, Deserialize, IcStorage)]
-pub struct FactoryState {
-    pub configuration: FactoryConfiguration,
-    pub factory: Factory,
+    pub fn version(&self) -> u32 {
+        self.version
+    }
 }
 
 impl Versioned for FactoryState {
-    type Previous = ();
+    type Previous = FactoryStateV1;
 
-    fn upgrade((): ()) -> Self {
-        Self::default()
-    }
-}
+    fn upgrade(prev: Self::Previous) -> Self {
+        let FactoryStateV1 {
+            configuration,
+            factory,
+        } = prev;
+        let Factory {
+            canisters,
+            checksum,
+        } = factory;
+
+        let hash = checksum.hash;
 
-impl Default for FactoryState {
-    fn default() -> Self {
         Self {
-            configuration: FactoryConfiguration::default(),
-            factory: Factory::new(&[]),
+            configuration,
+
+            // After the upgrade the canister wasm module would have to be uploaded again to
+            // provide the state header.
+            upgrading_module: None,
+
+            // We assume for now that the canisters were not modified by external controllers, as
+            // we didn't keep track of each canister has before.
+            canisters: canisters
+                .into_iter()
+                .map(|(principal, _)| (principal, hash.clone()))
+                .collect(),
+
+            update_lock: UpdateLock::default(),
         }
     }
 }
 
-/// This trait must be implemented by a factory state to make using of `init_factory_api` macro
-/// possible.
 impl FactoryState {
-    pub fn ledger_principal(&self) -> Principal {
-        self.configuration.ledger_principal
+    pub fn new(configuration: FactoryConfiguration) -> Self {
+        Self {
+            configuration,
+            ..Default::default()
+        }
+    }
+
+    /// Checks if the request caller is the factory conctroller (owner).
+    ///
+    /// # Errors
+    ///
+    /// Returns `FactoryError::AccessDenied` if the caller is not the factory controller.
+    pub fn authorize_owner(&mut self) -> Result<Authorized<Owner>, FactoryError> {
+        let caller = ic_canister::ic_kit::ic::caller();
+        self.authorize_owner_int(caller)
     }
 
+    // In some cases we need to call owner-only operations after `await` point, when `ic::caller()`
+    // cannot be called (due to IC limitation). In this case we use this method with the caller,
+    // retrieved before the `await` point. To prevent incorrect usage of this method it is
+    // `pub(crate)` only.
+    pub(crate) fn authorize_owner_int<'a>(
+        &'a mut self,
+        caller: Principal,
+    ) -> Result<Authorized<'a, Owner>, FactoryError> {
+        if caller == self.configuration.controller {
+            Ok(Authorized::<'a, Owner> {
+                factory: self,
+                phantom: PhantomData::default(),
+            })
+        } else {
+            Err(FactoryError::AccessDenied)
+        }
+    }
```

Comment:
Remove the life time specified here, it's not required:

```suggestion
    pub(crate) fn authorize_owner_int(
        &mut self,
        caller: Principal,
    ) -> Result<Authorized<'_, Owner>, FactoryError> {
        if caller == self.configuration.controller {
            Ok(Authorized::<'_, Owner> {
                factory: self,
                phantom: PhantomData::default(),
            })
        } else {
            Err(FactoryError::AccessDenied)
        }
    }
```